### PR TITLE
Initial torchbind support in PT2

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117697

This PR adds the bare minimum functionality to get torchbind working in an e2e testable way on PT2.

It implements:
* ProxyTensor support
* Simple torch.export support (proxytensor-only path, e.g. non-strict).
* add some tests exercising the path.

Because all this is not fully baked, I hide the functionality behind a feature flag (`enable_torchbind_tracing()`) so it does not affect regular users for now.

Still on the agenda:
* Dynamo support
* Actual FakeMode support
* Mutability support

Hoping to get this first bit in as a standalone, as it will unblock some more extensive experimentation/testing going on internally.

Differential Revision: [D51825372](https://our.internmc.facebook.com/intern/diff/D51825372/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by suo** [comment]
don't review the serialization bits, that will all get rebased away once @angelayi lands https://github.com/pytorch/pytorch/pull/116985

File: test/cpp/jit/test_custom_class_registrations.cpp [hunk 1]
```diff
@@ -436,4 +449,15 @@ TORCH_LIBRARY(_TorchScriptTesting, m) {
...
 
+at::Tensor takes_foo(c10::intrusive_ptr<Foo> foo, at::Tensor x) {
+  return foo->add_tensor(x);
+}
+
+TORCH_LIBRARY_IMPL(_TorchScriptTesting, CPU, m) {
+  m.impl("takes_foo", takes_foo);
+}
+TORCH_LIBRARY_IMPL(_TorchScriptTesting, Meta, m) {
+  m.impl("takes_foo", &takes_foo);
```
Comment (ezyang) [inline test/cpp/jit/test_custom_class_registrations.cpp:460]: spicy

File: torch/_higher_order_ops/torchbind.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+from torch._C import DispatchKey  # @manual
+from torch._functorch._aot_autograd.utils import KNOWN_TYPES
+from torch._higher_order_ops.utils import autograd_not_implemented
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.fx.node import has_side_effect
+from torch.utils import _pytree as pytree
+
+call_torchbind = HigherOrderOperator("call_torchbind")
```
Comment (ezyang) [inline torch/_higher_order_ops/torchbind.py:13]: Want a high level description about the op

**Comment by ezyang** [comment]
Yeah! This looks pretty plausible

**Review by SherlockNoMad — APPROVED** [review]

**Comment by suo** [comment]
Using `suppress-api-compatibility-check` as the affected API (`fetch_tensor_proxy`) is in an internal module (`torch.fx.experimental.proxy_tensor`)

**Pull request closed without merging**